File: questions/23.MergeKSortedLists.go
package questions

func mergeKLists(lists []*ListNode) *ListNode {
	if len(lists) == 0 {
		return nil
	}

	var merge func(list1, list2 *ListNode) *ListNode

	merge = func(list1, list2 *ListNode) *ListNode {
		head := &ListNode{}
		current := head

		for list1 != nil && list2 != nil {
			if list1.Val < list2.Val {
				current.Next = list1
				list1 = list1.Next
			} else {
				current.Next = list2
				list2 = list2.Next
			}
			current = current.Next
		}
		if list1 != nil {
			current.Next = list1
		}
		if list2 != nil {
			current.Next = list2
		}
		return head.Next
	}

	for len(lists) > 1 {
		var merged []*ListNode

		for i := 0; i < len(lists); i += 2 {
			list1 := lists[i]
			var list2 *ListNode
			if i+1 < len(lists) {
				list2 = lists[i+1]
			}

			merged = append(merged, merge(list1, list2))
		}
		lists = merged
	}

	return lists[0]
}

File: leetcode/23.MergeKSortedLists.ts
import { ListNode } from './shared'

function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
  if (!lists.length) return null

  const merge = (list1: ListNode | null, list2: ListNode | null) => {
    const head = new ListNode()
    let current = head

    while (list1 && list2) {
      if (list1.val < list2.val) {
        current.next = list1
        list1 = list1.next
      } else {
        current.next = list2
        list2 = list2.next
      }

      current = current.next
    }
    if (list1) {
      current.next = list1
    }
    if (list2) {
      current.next = list2
    }

    return head.next
  }

  while (lists.length > 1) {
    const merged = []
    for (let i = 0; i < lists.length; i += 2) {
      const list1 = lists[i]
      const list2 = lists?.[i + 1] ?? null
      merged.push(merge(list1, list2))
    }
    lists = merged
  }

  return lists[0]
}

const list1 = new ListNode(1, new ListNode(4, new ListNode(5, null)))
const list2 = new ListNode(1, new ListNode(3, new ListNode(4, null)))
const list3 = new ListNode(2, new ListNode(6, null))

console.log(mergeKLists([list1, list2, list3]))
